### PR TITLE
Support self-signed certificate and optional machine token

### PR DESCRIPTION
**Pull request by vzhukovs**
This changes proposal adds support self-signed certificates for activity track requests. That allow run prevent idling script on environments like CRW. Also added optional support of machine token which can be omitted on environment like minikube.

To check the script it is enough to run it on that environment where workspace actually stops after certain period of time.

Fixes https://github.com/eclipse/che/issues/18171

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### CODE REVIEW

**Review by l0rd — APPROVED** [review]

File: scripts/prevent-idle-timeout.sh [hunk 1]
```diff
@@ -8,27 +8,30 @@
...
-# Todo(vzhukovs): check for client certificates that might be included into request in case if script runs on minikube with self-signed certificates
+CURL="$CURL $CHE_API_INTERNAL/activity/$CHE_WORKSPACE_ID"
 
-# Start making requests to avoid idling, make a request to activity tracker each minute
 while true; do
-    curl -X PUT \
-    -H "Authorization: Bearer ${CHE_MACHINE_TOKEN}" \
-    "${CHE_API_INTERNAL}/activity/${CHE_WORKSPACE_ID}"
+	echo "$(date) - Perform request: '$CURL'."
+	eval "$CURL"
```
Comment (benoitf) [inline scripts/prevent-idle-timeout.sh:33]: I think it's less readable than having
CURL_OPTIONS (and adding --cacert) and authorization bearer

and still doing
```
$ curl -X PUT ${CURL_OPTIONS} "${CHE_API_INTERNAL}/activity/${CHE_WORKSPACE_ID}"
```

**Review by benoitf — APPROVED** [review]

File: scripts/prevent-idle-timeout.sh (hunk 1 above)
Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh:33]: I've tried this approach, but adding `set -x` before curl execution shows:
```
+ curl -X PUT --cacert /tmp/che/secret/ca.crt -H '"Authorization:' Bearer 'token"' http://example.com/activity/wsId
curl: (6) Could not resolve host: Bearer
curl: (6) Could not resolve host: token"
```


<details>
<summary>with this sample of script</summary>

```bash
#!/bin/sh
#
# Copyright (c) 2020 Red Hat, Inc.
# This program and the accompanying materials are made
# available under the terms of the Eclipse Public License 2.0
# which is available at https://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
#

CA_CERT="/tmp/che/secret/ca.crt"


if [ -z "${CHE_API_INTERNAL}" ] || [ -z "${CHE_WORKSPACE_ID}" ]; then
    echo "CHE_API_INTERNAL or CHE_WORKSPACE_ID is not configured in environment."
    exit 1
fi

CURL_OPTS=

if [ -e $CA_CERT ]; then
	CURL_OPTS="$CURL_OPTS"' --cacert /tmp/che/secret/ca.crt'
fi

if [ -n "$CHE_MACHINE_TOKEN" ]; then
	CURL_OPTS="$CURL_OPTS"' -H "Authorization: Bearer '"$CHE_MACHINE_TOKEN"'"'
fi

REQUEST_URL="$CHE_API_INTERNAL"'/activity/'"$CHE_WORKSPACE_ID"

while true; do
	echo "$(date) - Perform request: '$CURL_OPTS'."
	set -x
	curl -X PUT $CURL_OPTS $REQUEST_URL
	echo "$(date) - Sleep for 1 minute."

    sleep 60
done
```

</details>

and:
```
+ curl -X PUT ' --cacert /tmp/che/secret/ca.crt -H "Authorization: Bearer token"' http://example.com/activity/wsId
curl: (3) URL using bad/illegal format or missing URL
```

<details>
  <summary>when expanding curl opts</summary>
  
  ```bash
#!/bin/sh
#
# Copyright (c) 2020 Red Hat, Inc.
# This program and the accompanying materials are made
# available under the terms of the Eclipse Public License 2.0
# which is available at https://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
#

CA_CERT="/tmp/che/secret/ca.crt"


if [ -z "${CHE_API_INTERNAL}" ] || [ -z "${CHE_WORKSPACE_ID}" ]; then
    echo "CHE_API_INTERNAL or CHE_WORKSPACE_ID is not configured in environment."
    exit 1
fi

CURL_OPTS=

if [ -e $CA_CERT ]; then
	CURL_OPTS="$CURL_OPTS"' --cacert /tmp/che/secret/ca.crt'
fi

if [ -n "$CHE_MACHINE_TOKEN" ]; then
	CURL_OPTS="$CURL_OPTS"' -H "Authorization: Bearer '"$CHE_MACHINE_TOKEN"'"'
fi

REQUEST_URL="$CHE_API_INTERNAL"'/activity/'"$CHE_WORKSPACE_ID"

while true; do
	echo "$(date) - Perform request: '$CURL_OPTS'."
	set -x
	curl -X PUT "$CURL_OPTS" $REQUEST_URL
	echo "$(date) - Sleep for 1 minute."

    sleep 60
done

  ```
  
</details>

Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh:33]: But with original code I can say, that workspace on CRW 2.5 GA instance was started yesterday at ~10pm and it still active till now.

Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh:33]: Also PR was checked on che.openshift.io and minikube in the meantime.

**Review by azatsarynnyy — APPROVED** [review]
I'm approving based on the verifying results https://github.com/che-incubator/jetbrains-editor-images/pull/15#discussion_r518712947

File: scripts/prevent-idle-timeout.sh (hunk 1 above)
Comment (benoitf) [reply scripts/prevent-idle-timeout.sh:33]: @vzhukovs I think you're just missing the eval on curl command

```bash
#!/bin/bash
#
# Copyright (c) 2020 Red Hat, Inc.
# This program and the accompanying materials are made
# available under the terms of the Eclipse Public License 2.0
# which is available at https://www.eclipse.org/legal/epl-2.0/
#
# SPDX-License-Identifier: EPL-2.0
#

CA_CERT="/tmp/che/secret/ca.crt"


if [ -z "${CHE_API_INTERNAL}" ] || [ -z "${CHE_WORKSPACE_ID}" ]; then
    echo "CHE_API_INTERNAL or CHE_WORKSPACE_ID is not configured in environment."
    exit 1
fi

CURL_OPTS=""

if [ -e "${CA_CERT}" ]; then
  CURL_OPTS="$CURL_OPTS --cacert /tmp/che/secret/ca.crt"
fi

if [ -n "$CHE_MACHINE_TOKEN" ]; then
  CURL_OPTS="$CURL_OPTS -H \"Authorization: Bearer ${CHE_MACHINE_TOKEN}\""
fi

REQUEST_URL="${CHE_API_INTERNAL}/activity/${CHE_WORKSPACE_ID}"

while true; do
  echo "$(date) - Perform request: '$CURL_OPTS'."
  set -x
  eval curl -X PUT "${CURL_OPTS}" "${REQUEST_URL}"
  echo "$(date) - Sleep for 1 minute."
  sleep 60
done

```

Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh]: Much appreciated for your help! Now it works!

File: scripts/prevent-idle-timeout.sh [hunk 2]
```diff
@@ -8,27 +8,30 @@
...
 while true; do
-    curl -X PUT \
-    -H "Authorization: Bearer ${CHE_MACHINE_TOKEN}" \
-    "${CHE_API_INTERNAL}/activity/${CHE_WORKSPACE_ID}"
+	echo "$(date) - Perform request: 'curl -X PUT $CURL_OPTS $REQUEST_URL'."
+	eval curl -X PUT "${CURL_OPTS}" "${REQUEST_URL}"
+	echo "$(date) - Sleep for 1 minute."
 
-    sleep 1m
+    sleep 60
```
Comment (benoitf) [inline scripts/prevent-idle-timeout.sh:36]: seems that now you like tab characters instead of spaces

File: scripts/prevent-idle-timeout.sh [hunk 3]
```diff
@@ -8,27 +8,30 @@
...
     exit 1
 fi
-if [ -z "${CHE_MACHINE_TOKEN}" ]; then
-    echo "CHE_MACHINE_TOKEN is not configured in environment."
-    exit 1
+
+CURL_OPTS=
+
+if [ -e $CA_CERT ]; then
+	CURL_OPTS="$CURL_OPTS"' --cacert /tmp/che/secret/ca.crt'
```
Comment (benoitf) [inline scripts/prevent-idle-timeout.sh:22]:
```suggestion
	CURL_OPTS="$CURL_OPTS --cacert /tmp/che/secret/ca.crt"
```

File: scripts/prevent-idle-timeout.sh [hunk 4]
```diff
@@ -8,27 +8,30 @@
...
 fi
-if [ -z "${CHE_WORKSPACE_ID}" ]; then
-    echo "CHE_WORKSPACE_ID is not configured in environment."
-    exit 1
+
+if [ -n "$CHE_MACHINE_TOKEN" ]; then
+	CURL_OPTS="$CURL_OPTS"' -H "Authorization: Bearer '"$CHE_MACHINE_TOKEN"'"'
```
Comment (benoitf) [inline scripts/prevent-idle-timeout.sh:26]:
```suggestion
	CURL_OPTS="$CURL_OPTS -H \"Authorization: Bearer ${CHE_MACHINE_TOKEN}\""
```

File: scripts/prevent-idle-timeout.sh [hunk 5]
```diff
@@ -8,27 +8,30 @@
...
 fi
 
-# Todo(vzhukovs): check for client certificates that might be included into request in case if script runs on minikube with self-signed certificates
+REQUEST_URL="$CHE_API_INTERNAL"'/activity/'"$CHE_WORKSPACE_ID"
```
Comment (benoitf) [inline scripts/prevent-idle-timeout.sh:29]:
```suggestion
REQUEST_URL="${CHE_API_INTERNAL}/activity/${CHE_WORKSPACE_ID}"
```

File: scripts/prevent-idle-timeout.sh (hunk 2 above)
Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh:36]: Yeah, didn't noticed that. Just used a different editor. Updated

File: scripts/prevent-idle-timeout.sh (hunk 3 above)
Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh]: Updated

File: scripts/prevent-idle-timeout.sh (hunk 4 above)
Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh]: Updated

File: scripts/prevent-idle-timeout.sh (hunk 5 above)
Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh]: Updated

**Review by benoitf — APPROVED** [review]
approving again

File: scripts/prevent-idle-timeout.sh (hunk 2 above)
Comment (nickboldt) [reply scripts/prevent-idle-timeout.sh:36]: fwiw you can suppress whitespace changes when reviewing a PR with `?w=1` suffix, eg., https://github.com/che-incubator/jetbrains-editor-images/pull/15/files?w=1

Comment (vzhukovs) [reply scripts/prevent-idle-timeout.sh:36]: @nickboldt good to know, thanks, but agree, tab are evil

Comment (nickboldt) [reply scripts/prevent-idle-timeout.sh:36]: Re: tabs v. spaces:
https://dev.to/alexandersandberg/why-we-should-default-to-tabs-instead-of-spaces-for-an-accessible-first-environment-101f

If we're going to change our language to be more inclusive (master/slave/blacklist etc.) we should also not fear being more inclusive with tabs. Just sayin'... they're not evil.

**Pull request merged**